Sources/Datadog/RUM/RUMVitals/VitalInfoSampler.swift

```diff
@@ -0,0 +1,92 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import class UIKit.UIScreen
+
+internal protocol SamplingBasedVitalReader {
+    func readVitalData() -> Double?
+}
+
+internal protocol ContinuousVitalReader {
+    func register(_ valuePublisher: VitalPublisher)
+    func unregister(_ valuePublisher: VitalPublisher)
+}
+
+internal final class VitalInfoSampler {
+    private static let frequency: TimeInterval = 1.0
+
+    let cpuReader: SamplingBasedVitalReader
+    private let cpuPublisher = VitalPublisher(initialValue: VitalInfo())
+
+    var cpu: VitalInfo {
+        return cpuPublisher.currentValue
+    }
+
+    let memoryReader: SamplingBasedVitalReader
+    private let memoryPublisher = VitalPublisher(initialValue: VitalInfo())
+
+    var memory: VitalInfo {
+        return memoryPublisher.currentValue
+    }
+
+    let refreshRateReader: ContinuousVitalReader
+    private let refreshRatePublisher = VitalPublisher(initialValue: VitalInfo())
+    private let maximumRefreshRate: Double
+
+    var refreshRate: VitalInfo {
+        let info = refreshRatePublisher.currentValue
+        return info.scaledDown(by: maximumRefreshRate)
+    }
+
+    private var timer: Timer?
+
+    init(
+        cpuReader: SamplingBasedVitalReader,
+        memoryReader: SamplingBasedVitalReader,
+        refreshRateReader: ContinuousVitalReader,
+        frequency: TimeInterval = VitalInfoSampler.frequency,
+        maximumRefreshRate: Double = Double(UIScreen.main.maximumFramesPerSecond)
+    ) {
+        self.cpuReader = cpuReader
+        self.memoryReader = memoryReader
+        self.refreshRateReader = refreshRateReader
+        self.refreshRateReader.register(self.refreshRatePublisher)
+        self.maximumRefreshRate = maximumRefreshRate
+
+        takeSample()
+        let timer = Timer(
+            timeInterval: frequency,
+            target: self,
+            selector: #selector(takeSample),
+            userInfo: nil,
+            repeats: true
+        )
+        // NOTE: RUMM-1280 based on my running Example app
+        // non-main run loops don't fire the timer.
+        // Although i can't catch this in unit tests
```

Comment:
What can't be observed in unit tests? I'm not sure if I understand this comment - do we want to keep it?

Reply:
i added [this unit test](https://github.com/DataDog/dd-sdk-ios/pull/530/files#diff-c783872e2fe3a78e3fddd755e1d35685db2474c6df444818b0d4456eab6e4fcbR49) to catch cases where sampler is created in background thread so its timer wouldn't fire.
however this test never fails, regardless using `RunLoop.main` or `RunLoop.current` 🤔 

although when i run Example app with `RunLoop.current`, i can observe that the timer isn't fired

i think we should keep a comment here, the wording can be different